File: acgs2-core/sdk/python/examples/policy_registry_example.py
#!/usr/bin/env python3
"""
ACGS-2 Python SDK - Policy Registry Example
Constitutional Hash: cdd01ef066bc6cf2

This example demonstrates how to use the Policy Registry service
to manage policies, bundles, and authentication.
"""

import asyncio
import os

from acgs2_sdk import ACGS2Config, create_client


async def main():
    # Configuration
    config = ACGS2Config(
        base_url="http://localhost:8080",  # API Gateway URL
        tenant_id="example-tenant",
        # Add authentication as needed
        # api_key="your-api-key",
        # svid_token="your-svid-token",
    )

    async with create_client(config) as client:
        print("🛡️  ACGS-2 Policy Registry Example")
        print("=" * 50)

        # Get Policy Registry service
        policy_service = client.policy_registry

        try:
            # 1. Health Check
            print("\n1. Health Check")
            health = await policy_service.health_check()
            print(f"   Status: {health}")

            # 2. List Policies
            print("\n2. List Policies")
            policies = await policy_service.list_policies(limit=5)
            print(f"   Found {len(policies)} policies")
            if policies:
                policy = policies[0]
                print(f"   Example: {policy.name} (Status: {policy.status.value})")

            # 3. Create a Policy
            print("\n3. Create Policy")
            new_policy = await policy_service.create_policy(
                name="example-security-policy",
                rules=[
                    {
                        "effect": "allow",
                        "principal": "user:*",
                        "action": "read",
                        "resource": "document:*",
                        "conditions": {
                            "ip_address": {"type": "CIDR", "value": "192.168.1.0/24"}
                        }
                    }
                ],
                description="Example security policy with IP restrictions",
                tags=["security", "example"],
                compliance_tags=["gdpr", "sox"]
            )
            print(f"   Created policy: {new_policy.name} (ID: {new_policy.id})")

            # 4. Get Policy Details
            print("\n4. Get Policy Details")
            policy_details = await policy_service.get_policy(new_policy.id)
            print(f"   Policy: {policy_details.name}")
            print(f"   Rules: {len(policy_details.rules)}")
            print(f"   Status: {policy_details.status.value}")

            # 5. Verify Policy
            print("\n5. Verify Policy")
            verification = await policy_service.verify_policy(
                new_policy.id,
                {
                    "principal": "user:alice",
                    "action": "read",
                    "resource": "document:confidential",
                    "context": {
                        "ip_address": "192.168.1.100",
                        "time": "2024-01-15T10:00:00Z"
                    }
                }
            )
            print(f"   Verification: {'ALLOWED' if verification.allowed else 'DENIED'}")

            # 6. Get Policy Versions
            print("\n6. Get Policy Versions")
            versions = await policy_service.get_policy_versions(new_policy.id)
            print(f"   Policy has {len(versions)} versions")

            # 7. Create Policy Version
            print("\n7. Create Policy Version")
            new_version = await policy_service.create_policy_version(
                new_policy.id,
                rules=[
                    {
                        "effect": "allow",
                        "principal": "user:*",
                        "action": ["read", "write"],
                        "resource": "document:*",
                        "conditions": {
                            "ip_address": {"type": "CIDR", "value": "192.168.1.0/24"},
                            "department": {"type": "StringEquals", "value": "engineering"}
                        }
                    }
                ],
                description="Enhanced policy with write permissions and department restrictions"
            )
            print(f"   Created version: {new_version.version}")

            # 8. List Policy Bundles
            print("\n8. List Policy Bundles")
            bundles = await policy_service.list_bundles()
            print(f"   Found {len(bundles)} bundles")

            # 9. Create Policy Bundle
            print("\n9. Create Policy Bundle")
            bundle = await policy_service.create_bundle(
                name="security-bundle",
                policies=[new_policy.id],
                description="Bundle containing security policies"
            )
            print(f"   Created bundle: {bundle.name} (ID: {bundle.id})")

            # 10. Get Active Bundle
            print("\n10. Get Active Bundle")
            active_bundle = await policy_service.get_active_bundle()
            print(f"   Active bundle: {active_bundle.name}")

            # 11. Authentication (if credentials provided)
            if os.getenv("ACGS2_USERNAME") and os.getenv("ACGS2_PASSWORD"):
                print("\n11. Authentication")
                auth_result = await policy_service.authenticate(
                    os.getenv("ACGS2_USERNAME"),
                    os.getenv("ACGS2_PASSWORD")
                )
                print(f"   Authenticated as: {auth_result.user.username}")
                print(f"   Roles: {', '.join(auth_result.user.roles)}")

            print("\n✅ Policy Registry example completed successfully!")

        except Exception as e:
            print(f"\n❌ Example failed: {e}")
            import traceback
            traceback.print_exc()


if __name__ == "__main__":
    asyncio.run(main())

File: acgs2-core/sdk/typescript/examples/policy-registry-example.ts
/**
 * ACGS-2 TypeScript SDK - Policy Registry Example
 * Constitutional Hash: cdd01ef066bc6cf2
 *
 * This example demonstrates how to use the Policy Registry service
 * to manage policies, bundles, and authentication.
 */

import { ACGS2Client } from '../src/client';

async function main() {
  console.log('🛡️  ACGS-2 Policy Registry Example');
  console.log('='.repeat(50));

  // Configuration
  const client = new ACGS2Client({
    baseURL: 'http://localhost:8080', // API Gateway URL
    tenantId: 'example-tenant',
    // Add authentication as needed
    // apiKey: 'your-api-key',
    // svidToken: 'your-svid-token',
  });

  try {
    const policyService = client.policyRegistry();

    // 1. Health Check
    console.log('\n1. Health Check');
    const health = await policyService.healthCheck();
    console.log(`   Status:`, health);

    // 2. List Policies
    console.log('\n2. List Policies');
    const policies = await policyService.listPolicies({ limit: 5 });
    console.log(`   Found ${policies.length} policies`);
    if (policies.length > 0) {
      const policy = policies[0];
      console.log(`   Example: ${policy.name} (Status: ${policy.status})`);
    }

    // 3. Create a Policy
    console.log('\n3. Create Policy');
    const newPolicy = await policyService.createPolicy({
      name: 'example-security-policy',
      rules: [
        {
          effect: 'allow',
          principal: 'user:*',
          action: 'read',
          resource: 'document:*',
          conditions: {
            ip_address: { type: 'CIDR', value: '192.168.1.0/24' }
          }
        }
      ],
      description: 'Example security policy with IP restrictions',
      tags: ['security', 'example'],
      complianceTags: ['gdpr', 'sox']
    });
    console.log(`   Created policy: ${newPolicy.name} (ID: ${newPolicy.id})`);

    // 4. Get Policy Details
    console.log('\n4. Get Policy Details');
    const policyDetails = await policyService.getPolicy(newPolicy.id);
    console.log(`   Policy: ${policyDetails.name}`);
    console.log(`   Rules: ${policyDetails.rules.length}`);
    console.log(`   Status: ${policyDetails.status}`);

    // 5. Verify Policy
    console.log('\n5. Verify Policy');
    const verification = await policyService.verifyPolicy(newPolicy.id, {
      input: {
        principal: 'user:alice',
        action: 'read',
        resource: 'document:confidential',
        context: {
          ip_address: '192.168.1.100',
          time: '2024-01-15T10:00:00Z'
        }
      }
    });
    console.log(`   Verification: ${verification.allowed ? 'ALLOWED' : 'DENIED'}`);
    if (!verification.allowed && verification.reason) {
      console.log(`   Reason: ${verification.reason}`);
    }

    // 6. Get Policy Versions
    console.log('\n6. Get Policy Versions');
    const versions = await policyService.getPolicyVersions(newPolicy.id);
    console.log(`   Policy has ${versions.length} versions`);

    // 7. Create Policy Version
    console.log('\n7. Create Policy Version');
    const newVersion = await policyService.createPolicyVersion(
      newPolicy.id,
      {
        rules: [
          {
            effect: 'allow',
            principal: 'user:*',
            action: ['read', 'write'],
            resource: 'document:*',
            conditions: {
              ip_address: { type: 'CIDR', value: '192.168.1.0/24' },
              department: { type: 'StringEquals', value: 'engineering' }
            }
          }
        ]
      },
      'Enhanced policy with write permissions and department restrictions'
    );
    console.log(`   Created version: ${newVersion.version}`);

    // 8. List Policy Bundles
    console.log('\n8. List Policy Bundles');
    const bundles = await policyService.listBundles();
    console.log(`   Found ${bundles.length} bundles`);

    // 9. Create Policy Bundle
    console.log('\n9. Create Policy Bundle');
    const bundle = await policyService.createBundle({
      name: 'security-bundle',
      policies: [newPolicy.id],
      description: 'Bundle containing security policies'
    });
    console.log(`   Created bundle: ${bundle.name} (ID: ${bundle.id})`);

    // 10. Get Active Bundle
    console.log('\n10. Get Active Bundle');
    const activeBundle = await policyService.getActiveBundle();
    console.log(`   Active bundle: ${activeBundle.name}`);

    // 11. Authentication (if credentials provided)
    const username = process.env.ACGS2_USERNAME;
    const password = process.env.ACGS2_PASSWORD;

    if (username && password) {
      console.log('\n11. Authentication');
      const authResult = await policyService.authenticate({
        username,
        password
      });
      console.log(`   Authenticated as: ${authResult.user.username}`);
      console.log(`   Roles: ${authResult.user.roles.join(', ')}`);
    }

    console.log('\n✅ Policy Registry example completed successfully!');

  } catch (error) {
    console.error('\n❌ Example failed:', error);
    throw error;
  } finally {
    client.close();
  }
}

// Run the example
main().catch(console.error);
